unlock-js/src/unlockService.js

```diff
@@ -136,4 +151,132 @@ export default class UnlockService extends EventEmitter {
+    }
+
+    let version = this.ethers_versionForAddress[this.unlockContractAddress]
+    if (version === undefined) {
+      // This was not memo-ized
+      version = await this._ethers_getVersionFromContract(
+        this.unlockContractAddress
+      )
+      this.ethers_versionForAddress[this.unlockContractAddress] = version
+    }
+
+    if (1 === version) {
+      return v01
+    }
+
+    if (2 === version) {
+      return v02
+    }
+
+    // Defaults to v0
+    return v0
+  }
+
+  /**
+   * Returns the ABI for the Lock contract deployed at the provided address
+   * @param {*} address
+   */
+  async ethers_lockContractAbiVersion(address) {
+    if (!this.provider) {
+      throw new Error(Errors.MISSING_WEB3)
+    }
+
+    let version = this.ethers_versionForAddress[address.toLowerCase()]
+    if (version === undefined) {
+      // This was not memo-ized
+      version = await this._ethers_getPublicLockVersionFromContract(address)
+      this.ethers_versionForAddress[address.toLowerCase()] = version
+    }
+
+    // NOTE: we (julien) F'ed up the deploy on the PublicLock and v02 still uses 1 for its version.
+    // The good news (luck) is that no contract was ever deployed as v01
+    if (1 === version) {
+      return v02
+    }
+
+    // Defaults to v0
+    return v0
+  }
+
+  async _ethers_getPublicLockVersionFromContract(address) {
+    const contract = new Contract(
+      address,
+      ['function publicLockVersion() view returns (uint8)'],
+      this.provider
+    )
+    let version = 0
+    try {
+      const contractVersion = await contract.publicLockVersion()
+      version = parseInt(contractVersion, 10) || 0
+    } catch (error) {
+      // This is an older version of Unlock which did not support publicLockVersion
+    }
+    return version
+  }
+
+  /**
+   * Private method, which given an address will query the contract and return the corresponding method
+   * @param {*} address
+   */
+  async _ethers_getVersionFromContract(address) {
+    const contract = new Contract(
+      address,
+      ['function unlockVersion() view returns (uint8)'],
+      this.provider
+    )
+    let version = 0
+    try {
+      const contractVersion = await contract.unlockVersion()
+      version = parseInt(contractVersion, 10) || 0
+    } catch (error) {
+      // This is an older version of Unlock which did not support unlockVersion
+    }
+    return version
+  }
+
+  getContract(address, contract) {
+    if (this.writable) return this.getWritableContract(address, contract)
+    return new Contract(address, contract.abi, this.provider)
+  }
+
+  async getWritableContract(address, contract) {
+    const signer = this.provider.getSigner()
+    return new Contract(address, contract.abi, signer)
+  }
+
+  async getLockContract(lockAddress) {
+    if (this.lockContracts[lockAddress]) {
+      return this.lockContracts[lockAddress]
+    }
+    const version = await this.ethers_lockContractAbiVersion(lockAddress)
```

Comment:
Would you be against the remaining portion of this in an else?

Reply:
done